server/database.go

```diff
@@ -959,7 +1081,59 @@ func getDatabaseDriver(ctx context.Context, instance *api.Instance, databaseName
 		},
 	)
 	if err != nil {
-		return nil, common.Errorf(common.DbConnectionFailure, fmt.Errorf("failed to connect database at %s:%s with user %q: %w", instance.Host, instance.Port, instance.Username, err))
+		return nil, err
+	}
+
+	return driver, nil
+}
+
+// Try to get database driver using the instance's read-only data source.
+// If no read-only data source is found, try using the admin data source.
+// Upon successful return, caller MUST call driver.Close, otherwise, it will leak the database connection.
+func tryGetReadOnlyDatabaseDriver(ctx context.Context, instance *api.Instance, databaseName string, logger *zap.Logger) (db.Driver, error) {
+	readOnlyDataSource := api.DataSourceFromInstanceWithType(instance, api.RO)
```

Comment:
This should be called dataSource since it can be adminDataSource.

Also add a comment saying we'd like to use read-only data source whenever possible but fallback to admin data source if there's no read-only data source.